org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/ContentModelFormatterParticipant.java

```diff
@@ -67,4 +67,30 @@ public FormatElementCategory getFormatElementCategory(DOMElement element,
 		return null;
 	}
 
+	@Override
+	public boolean shouldCollapseEmptyElement(DOMElement element, SharedSettings sharedSettings) {
+		boolean enabled = sharedSettings.getFormattingSettings().isGrammarAwareFormatting();
+		if (!enabled) {
+			return true;
+		}
+		if (element.getAttribute("xsi:nil") != null) {
+			Collection<CMDocument> cmDocuments = contentModelManager.findCMDocument(element);
+			if (cmDocuments.isEmpty()) {
+				// The DOM document is not linked to a grammar, the collapse can be done.
+				return true;
+			}
+			for (CMDocument cmDocument : cmDocuments) {
+				CMElementDeclaration cmElement = cmDocument.findCMElement(element);
+				if (cmElement != null && cmElement.isEmpty()) {
+					return true;
+				} else if (cmElement != null
+						&& (!cmElement.isNillable() && "true".equals(element.getAttribute("xsi:nil")))) {

```

Comment:
As condition "true".equals(element.getAttribute("xsi:nil"))) is done below, you don't need to test it. 

cmElement.isNillable() returns false if XSD doesn't defines nillable attribute, is it correct?

Reply:
Yes it will return false because nillable is set to false by default.